File: libs/cua-driver/sdk/python/src/cua_driver/_expect.py
"""expect() — assertion helper for cua-driver Locators.

Modelled after Playwright's ``expect(locator).to_have_value(...)`` API.
Each assertion retries until the condition holds or the timeout elapses.
"""

from __future__ import annotations

import time
from typing import Optional

from ._locator import Locator


class LocatorAssertions:
    """Fluent assertion interface for a :class:`~cua_driver.Locator`.

    Obtain via :func:`expect`::

        expect(window.locator("AXStaticText")).to_contain_text("4")
    """

    def __init__(self, locator: Locator, timeout: float = 5.0) -> None:
        self._locator = locator
        self._timeout = timeout

    # ------------------------------------------------------------------
    # Positive assertions
    # ------------------------------------------------------------------

    def to_be_visible(self, timeout: Optional[float] = None) -> None:
        """Assert the element is present in the AX tree."""
        t = timeout if timeout is not None else self._timeout
        # _wait_for_element raises TimeoutError on failure, which is an
        # assertion error in a test context.
        self._locator._wait_for_element(t)

    def to_contain_text(self, text: str, timeout: Optional[float] = None) -> None:
        """Assert that at least one matching element's visible text contains *text*.

        Checks ALL elements returned by the locator — important when there are
        multiple elements of the same role (e.g. the Calculator result display
        and the expression-history display both being ``AXStaticText``).
        """
        t = timeout if timeout is not None else self._timeout
        deadline = time.monotonic() + t
        last_content = "<not found>"
        while True:
            contents = self._locator.all_text_contents()
            for content in contents:
                if text in content:
                    return
            if contents:
                last_content = repr(contents)
            if time.monotonic() >= deadline:
                raise AssertionError(
                    f"Expected element to contain text {text!r}, "
                    f"but got {last_content} after {t}s"
                )
            time.sleep(0.3)

    def to_have_text(self, text: str, timeout: Optional[float] = None) -> None:
        """Assert that at least one matching element's visible text equals *text* exactly."""
        t = timeout if timeout is not None else self._timeout
        deadline = time.monotonic() + t
        last_content = "<not found>"
        while True:
            contents = self._locator.all_text_contents()
            for content in contents:
                if content == text:
                    return
            if contents:
                last_content = repr(contents)
            if time.monotonic() >= deadline:
                raise AssertionError(
                    f"Expected element text to equal {text!r}, "
                    f"but got {last_content} after {t}s"
                )
            time.sleep(0.3)

    def to_have_value(self, value: str, timeout: Optional[float] = None) -> None:
        """Assert the element's ``value=`` attribute equals *value*."""
        t = timeout if timeout is not None else self._timeout
        deadline = time.monotonic() + t
        last_val = "<not found>"
        while True:
            try:
                val = self._locator.get_value(timeout=0.0)
                if val == value:
                    return
                last_val = val
            except TimeoutError:
                pass
            if time.monotonic() >= deadline:
                raise AssertionError(
                    f"Expected element value to equal {value!r}, "
                    f"but got {last_val!r} after {t}s"
                )
            time.sleep(0.3)

    # ------------------------------------------------------------------
    # Negative assertions
    # ------------------------------------------------------------------

    def not_to_be_visible(self, timeout: Optional[float] = None) -> None:
        """Assert the element is NOT present in the AX tree."""
        t = timeout if timeout is not None else self._timeout
        deadline = time.monotonic() + t
        while True:
            if not self._locator.is_visible(timeout=0.0):
                return
            if time.monotonic() >= deadline:
                raise AssertionError(
                    f"Expected element to not be visible, but it was still "
                    f"present after {t}s"
                )
            time.sleep(0.3)


def expect(locator: Locator, timeout: float = 5.0) -> LocatorAssertions:
    """Return a :class:`LocatorAssertions` for *locator*.

    Usage::

        from cua_driver import expect

        expect(window.locator("AXStaticText")).to_contain_text("4")
        expect(window.locator("AXButton", label="OK")).to_be_visible()
    """
    return LocatorAssertions(locator, timeout)

File: libs/cua-driver/sdk/typescript/src/expect.ts
/**
 * expect() — assertion helper for cua-driver Locators.
 *
 * Retries assertions until the condition holds or the timeout elapses.
 *
 * Usage:
 * ```ts
 * import { expect } from './expect.js'
 *
 * await expect(window.locator('AXStaticText')).toContainText('4')
 * await expect(window.locator('AXButton', { label: 'OK' })).toBeVisible()
 * ```
 */

import type { Locator } from './locator.js'

const POLL_INTERVAL = 300  // ms

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms))
}

export class LocatorAssertions {
  constructor(
    private readonly locator: Locator,
    private readonly timeoutMs: number,
  ) {}

  async toBeVisible(timeoutMs?: number): Promise<void> {
    const t = timeoutMs ?? this.timeoutMs
    const deadline = Date.now() + t
    while (true) {
      if (await this.locator.isVisible(0)) return
      if (Date.now() >= deadline) throw new Error(`Element not visible after ${t}ms`)
      await sleep(POLL_INTERVAL)
    }
  }

  async notToBeVisible(timeoutMs?: number): Promise<void> {
    const t = timeoutMs ?? this.timeoutMs
    const deadline = Date.now() + t
    while (true) {
      if (!await this.locator.isVisible(0)) return
      if (Date.now() >= deadline) throw new Error(`Element still visible after ${t}ms`)
      await sleep(POLL_INTERVAL)
    }
  }

  async toContainText(text: string, timeoutMs?: number): Promise<void> {
    const t = timeoutMs ?? this.timeoutMs
    const deadline = Date.now() + t
    let last = '<not found>'
    while (true) {
      // Check ALL matching elements — needed when multiple elements share a role
      // (e.g. Calculator result + expression-history are both AXStaticText).
      const contents = await this.locator.allTextContents()
      for (const content of contents) {
        if (content.includes(text)) return
      }
      if (contents.length) last = JSON.stringify(contents)
      if (Date.now() >= deadline) {
        throw new Error(`Expected text "${text}" not found in ${last} after ${t}ms`)
      }
      await sleep(POLL_INTERVAL)
    }
  }

  async toHaveText(text: string, timeoutMs?: number): Promise<void> {
    const t = timeoutMs ?? this.timeoutMs
    const deadline = Date.now() + t
    let last = '<not found>'
    while (true) {
      const contents = await this.locator.allTextContents()
      for (const content of contents) {
        if (content === text) return
      }
      if (contents.length) last = JSON.stringify(contents)
      if (Date.now() >= deadline) {
        throw new Error(`Expected text "${text}" but got ${last} after ${t}ms`)
      }
      await sleep(POLL_INTERVAL)
    }
  }

  async toHaveValue(value: string, timeoutMs?: number): Promise<void> {
    const t = timeoutMs ?? this.timeoutMs
    const deadline = Date.now() + t
    let last = '<not found>'
    while (true) {
      try {
        const val = await this.locator.getValue(0)
        if (val === value) return
        last = val
      } catch { /* element not found yet */ }
      if (Date.now() >= deadline) {
        throw new Error(`Expected value "${value}" but got "${last}" after ${t}ms`)
      }
      await sleep(POLL_INTERVAL)
    }
  }
}

/**
 * Return a {@link LocatorAssertions} for *locator*.
 *
 * @example
 * ```ts
 * await expect(window.locator('AXStaticText')).toContainText('4')
 * ```
 */
export function expect(locator: Locator, timeoutMs = 5_000): LocatorAssertions {
  return new LocatorAssertions(locator, timeoutMs)
}
